File: inquiry.py
import requests
from datetime import datetime
from bs4 import BeautifulSoup

def cj_inquiry(invoice):
    global db
    url = "http://nexs.cjgls.com/web"
    suffix = "/info.jsp"
    headers = { "Content-Type": "application/json;charset=UTF-8" }
    params = { 'slipno': str(invoice) }

    try:
        response = requests.get(url+suffix, headers=headers, params=params)
    except:
        return {'success': False, 'msg': 'request timesrror'}

    soup = BeautifulSoup(response.text, 'html5lib')
    items = soup.select('body > center > table:nth-child(9) > tbody > tr')
    inquiry = []
    inqLine = {}
    
    if items[1].select('td')[0].text == "해당 정보가 없습니다.":
        return {'success': False, 'msg': 'false invoice'}
    for item in items[1:]:
        temp = item.select('body > center > table:nth-child(9) > tbody > tr > td')
        inqLine['dateTime'] = int(datetime.strptime( temp[0].text.strip() + " " + temp[1].text.strip() , '%Y-%m-%d %H:%M:%S').timestamp())//60
        inqLine['loc'] = temp[2].text[:temp[2].text.index('Tel')].strip()
        if '(' in inqLine['loc']:
            inqLine['loc'] = inqLine['loc'][:inqLine['loc'].index('(')] + inqLine['loc'][inqLine['loc'].index(')')+1:]
        if '[' in inqLine['loc']:
            inqLine['loc'] = inqLine['loc'][:inqLine['loc'].index('[')] + inqLine['loc'][inqLine['loc'].index(']')+1:]
        inqLine['process'] = temp[3].text.strip()
        inquiry.append(inqLine.copy())

    if len(inquiry) <= 1:
        return {'success': False, 'msg': 'delivery is not finished'}

    result = []
    for line in inquiry[1:]:
        date = datetime.fromtimestamp(line['dateTime'] * 60)
        prog = f"{max(date.weekday()-3, 0)}T{date.hour // 2}_{line['loc']}_{line['process']}"

        result.append((prog, inquiry[0]['dateTime'] - line['dateTime']))
    result.append(('cj', result[-1][1]))

    return {'success': True, 'res': result}

def po_inquiry(invoice):
    global db
    url = "https://service.epost.go.kr"
    suffix = "/trace.RetrieveDomRigiTraceList.comm"
    headers = { "Content-Type": "application/json;charset=UTF-8" }
    params = { 'sid1': str(invoice) }

    try:
        response = requests.get(url+suffix, headers=headers, params=params)
    except:
        return {'success': False, 'msg': 'request timesrror'}

    soup = BeautifulSoup(response.text, 'html5lib')
    items = soup.select('body > #sbody_layout > #contents_layout > .contents > .h4_wrap > #print > :nth-child(6) > table > tbody > tr')

    inquiry = []
    inqLine = {}
    
    if len(items) == 0:
        return {'success': False, 'msg': 'false invoice'}

    for item in items:
        temp = item.select('td')
        
        inqLine['dateTime'] = int(datetime.strptime( temp[0].text.strip() + " " + temp[1].text.strip() , '%Y.%m.%d %H:%M').timestamp())//60
        inqLine['loc'] = temp[2].text.strip()
        for char in ['(', '[', 'TEL', '\n', '\t']:
            if char in inqLine['loc'].upper():
                inqLine['loc'] = inqLine['loc'][:inqLine['loc'].upper().index(char)]
        inqLine['process'] = temp[3].text.strip()
        for char in ['(', '[', 'TEL', '\n', '\t']:
            if char in inqLine['process'].upper():
                inqLine['process'] = inqLine['process'][:inqLine['process'].upper().index(char)]
        inquiry.append(inqLine.copy())

    if len(inquiry) <= 1:
        return {'success': False, 'msg': 'delivery is not finished'}

    result = []
    for line in inquiry[:-1]:
        date = datetime.fromtimestamp(line['dateTime'] * 60)
        prog = f"{max(date.weekday()-3, 0)}T{date.hour // 2}_{line['loc']}_{line['process']}"

        result.append((prog, inquiry[-1]['dateTime'] - line['dateTime']))
    result.append(('po', result[0][1]))

    return {'success': True, 'res': result}

def hj_inquiry(invoice):
    global db
    url = "https://www.hanjin.com"
    suffix = "/kor/CMS/DeliveryMgr/WaybillResult.do"
    headers = { "Content-Type": "application/json;charset=UTF-8" }
    params = { 'mCode': 'MN038', 'schLang': 'KR', 'wblnumText2': str(invoice) }

    try:
        response = requests.get(url+suffix, headers=headers, params=params)
    except:
        return {'success': False, 'msg': 'request timesrror'}

    soup = BeautifulSoup(response.text, 'html5lib')
    items = soup.select('body > #doc-wrap > #container-wrap > #container > #contents #cont > #delivery-wr > .cont-box > .waybill-tbl > table > tbody > tr')

    inquiry = []
    inqLine = {}
    
    if len(items) == 0:
        return {'success': False, 'msg': 'false invoice'}

    for item in items:
        temp = item.select('td')
        
        inqLine['dateTime'] = int(datetime.strptime( temp[0].text.strip() + " " + temp[1].text.strip() , '%Y-%m-%d %H:%M').timestamp())//60
        inqLine['loc'] = temp[2].text.strip()
        for char in ['(', '[', 'TEL', '\n', '\t']:
            if char in inqLine['loc'].upper():
                inqLine['loc'] = inqLine['loc'][:inqLine['loc'].upper().index(char)]
        
        tempS = temp[3].text
        if '입고' in tempS:
            inqLine['process'] = '입고'
        elif '접수' in tempS:
            inqLine['process'] = '접수'
        elif '준비' in tempS:
            inqLine['process'] = '배송준비'
        elif '출발' in tempS:
            inqLine['process'] = '배송출발'
        elif '완료' in tempS:
            inqLine['process'] = '배송완료'
        else:
            inqLine['process'] = '이동'#'간선상차'
        inquiry.append(inqLine.copy())
    if len(inquiry) <= 1:
        return {'success': False, 'msg': 'delivery is not finished'}

    result = []
    for line in inquiry[:-1]:
        date = datetime.fromtimestamp(line['dateTime'] * 60)
        prog = f"{max(date.weekday()-3, 0)}T{date.hour // 2}_{line['loc']}_{line['process']}"

        result.append((prog, inquiry[-1]['dateTime'] - line['dateTime']))
    result.append(('hj', result[0][1]))

    return {'success': True, 'res': result}

def lg_inquiry(invoice):
    global db
    url = "https://www.ilogen.com/"
    suffix = "/web/personal/trace/" + str(invoice)
    headers = { "Content-Type": "application/json;charset=UTF-8" }
    params = {  }

    try:
        response = requests.get(url+suffix, headers=headers, params=params)
    except:
        return {'success': False, 'msg': 'request timesrror'}

    soup = BeautifulSoup(response.text, 'html5lib')
    items = soup.select('body > div.contents.personal.tkSearch > section > div > div.tab_container > div > table.data.tkInfo > tbody > tr')

    inquiry = []
    inqLine = {}
    
    if len(items) == 0:
        return {'success': False, 'msg': 'false invoice'}
    for item in items:
        temp = item.select('td')
        
        inqLine['dateTime'] = int(datetime.strptime( temp[0].text.strip() , '%Y.%m.%d %H:%M').timestamp())//60
        inqLine['loc'] = temp[1].text.strip()
        inqLine['process'] = temp[2].text.strip()
        inquiry.append(inqLine.copy())
    if len(inquiry) <= 1:
        return {'success': False, 'msg': 'delivery is not finished'}

    result = []
    for line in inquiry[:-1]:
        date = datetime.fromtimestamp(line['dateTime'] * 60)
        prog = f"{max(date.weekday()-3, 0)}T{date.hour // 2}_{line['loc']}_{line['process']}"

        result.append((prog, inquiry[-1]['dateTime'] - line['dateTime']))
    result.append(('lg', result[0][1]))

    return {'success': True, 'res': result }

func = {'cj': cj_inquiry, 'po': po_inquiry, 'hj': hj_inquiry, 'lg': lg_inquiry}

if __name__ == "__main__":
    #print(f'{6892049692080}: {po_inquiry(6892049692080)}')
    print(f'{6892049692081}: {po_inquiry(6892049692081)}')

File: node_inquiry/Inquiry.js
"use strict";

const axios = require("axios");
const iconv = require('iconv-lite');

async function cj(invoice) {
    const html = await axios.get(`http://nplus.doortodoor.co.kr/web/detail.jsp?slipno=${invoice}`, {
        responseType: 'arraybuffer',
    });
    
    const content = iconv.decode(html.data, 'euc-kr').toString();
    const tableList = content.split('<tr')
    var inquiry = [];

    for (var i = 11;i+1 < tableList.length;i+=2){
        var timeList = tableList[i].split('&nbsp');
        var locList = tableList[i+1].split('&nbsp');

        inquiry.push({
            'location': locList[1].slice(1), 
            'status': locList[5].slice(1),
            'time': `${timeList[1].slice(1)} ${timeList[3].slice(1, -3)}`
        });
    }
    return inquiry;
}
async function po(invoice) {
    const html = await axios.get(`https://service.epost.go.kr/trace.RetrieveDomRigiTraceList.comm?sid1=${invoice}`, {
        responseType: 'arraybuffer',
    });
    
    const content = iconv.decode(html.data, 'utf-8').toString();
    const tableList = content.split('<tr')
    var inquiry = [];
    if (tableList.length <= 5){ return inquiry; }

    for (var i = 4;i < tableList.length;i++){
        var locStatus = tableList[i].split('<td>')[3].split('</')[0]
        
        inquiry.push({
            'location': locStatus.split('>')[locStatus.split('>').length-1], 
            'status': locStatus.split('\'')[1],
            'time': `${tableList[i].split('<td>')[1].split('</')[0].split('.').join('-')} ${tableList[i].split('<td>')[2].split('</')[0]}`
        });
    }
    inquiry.reverse();
    return inquiry;
}
async function lg(invoice) {
    const html = await axios.get(`https://www.ilogen.com/web/personal/trace/${invoice}`, {
        responseType: 'arraybuffer',
    });
    
    const content = iconv.decode(html.data, 'utf-8').toString();
    const tableList = content.split('<tr')
    var inquiry = []
    if (tableList.length <= 7){ return inquiry; }

    for (var i = 7;i < tableList.length-3;i++){
        const values = tableList[i].split('<td>').slice(1, 4);
        
        inquiry.push({
            'location': values[1].split('</')[0], 
            'status': values[2].split('\n')[0],
            'time': values[0].split('</')[0].split('.').join('-')
        });
    }
    inquiry.reverse();
    return inquiry;
}
async function hj(invoice) {
    const html = await axios.get(`https://www.hanjin.com/kor/CMS/DeliveryMgr/WaybillResult.do?mCode=MN038&schLang=KR&wblnumText2=${invoice}`, {
        responseType: 'arraybuffer',
    });
    
    const content = iconv.decode(html.data, 'utf-8').toString();
    const tableList = content.split('tbody>')[3].split('<tr');
    var inquiry = []

    for (var i = 1;i < tableList.length;i++){
        const values = tableList[i].split('<td').slice(1, 6);
        var status = '이동'
        if(values[3].includes('접수')) { status = '접수'; }
        else if (values[3].includes('입고')) { status = '입고'; }
        else if (values[3].includes('준비')) { status = '배송준비'; }
        else if (values[3].includes('출발')) { status = '배송출발'; }
        else if (values[3].includes('완료')) { status = '배송완료'; }
        
        inquiry.push({
            'location': values[2].slice(values[2].indexOf('>')+1, values[2].indexOf('</td>')), 
            'status': status,
            'time': values[0].slice(values[0].indexOf('>')+1, values[0].indexOf('</td>')) + ' ' + values[1].slice(values[1].indexOf('>')+1, values[1].indexOf('</td>'))
        });
    }
    inquiry.reverse();
    return inquiry;
}